### PR TITLE
New Intel: Slack

### PR DESCRIPTION
**Pull request by jychp**
### Summary

> Add intel to sync Slack Teams, Groups, Users and Channels

If you are modifying or implementing a new intel module
- [x] Update the [schema](https://github.com/lyft/cartography/tree/master/docs/root/modules) and [readme](https://github.com/lyft/cartography/blob/master/docs/schema/README.md)
- [X] Use our NodeSchema [data model](https://lyft.github.io/cartography/dev/writing-intel-modules.html#defining-a-node)
- [X] Use specialized functions `get_`, `transform_`, `load_`, and `cleanup_` functions
- [x] Add [tests](https://lyft.github.io/cartography/dev/writing-intel-modules.html#making-tests)
  - [ ] Unit tests: Test your `transform_` function with sample data
  - [x] Integration tests

### CODE REVIEW

File: cartography/models/slack/group.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    description: PropertyRef = PropertyRef('description')
+    is_subteam: PropertyRef = PropertyRef('is_subteam')
+    handle: PropertyRef = PropertyRef('handle')
+    is_external: PropertyRef = PropertyRef('is_external')
+    date_create: PropertyRef = PropertyRef('date_create')
+    date_update: PropertyRef = PropertyRef('date_update')
+    date_delete: PropertyRef = PropertyRef('date_delete')
+    updated_by: PropertyRef = PropertyRef('updated_by')
+    user_count: PropertyRef = PropertyRef('user_count')
+    channel_count: PropertyRef = PropertyRef('channel_count')
```
Comment (ramonpetgrave64) [inline cartography/models/slack/group.py:28]: Add the created_by property

File: cartography/models/slack/group.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+from cartography.models.core.relationships import make_target_node_matcher
+from cartography.models.core.relationships import OtherRelationships
+from cartography.models.core.relationships import TargetNodeMatcher
+
+
+@dataclass(frozen=True)
+class SlackGroupNodeProperties(CartographyNodeProperties):
+    id: PropertyRef = PropertyRef('id')
+    lastupdated: PropertyRef = PropertyRef('lastupdated', set_in_kwargs=True)
+    name: PropertyRef = PropertyRef('name')
```
Comment (ramonpetgrave64) [inline cartography/models/slack/group.py:18]: Could be worth an extra_index=True

File: cartography/models/slack/channels.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+from cartography.models.core.relationships import make_target_node_matcher
+from cartography.models.core.relationships import OtherRelationships
+from cartography.models.core.relationships import TargetNodeMatcher
+
+
+@dataclass(frozen=True)
+class SlackChannelNodeProperties(CartographyNodeProperties):
+    id: PropertyRef = PropertyRef('id')
+    lastupdated: PropertyRef = PropertyRef('lastupdated', set_in_kwargs=True)
+    name: PropertyRef = PropertyRef('name')
```
Comment (ramonpetgrave64) [inline cartography/models/slack/channels.py:18]: Could be worth an extra_index=True

File: cartography/models/slack/user.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    id: PropertyRef = PropertyRef('id')
+    lastupdated: PropertyRef = PropertyRef('lastupdated', set_in_kwargs=True)
+    name: PropertyRef = PropertyRef('name')
+    real_name: PropertyRef = PropertyRef('real_name')
+    display_name: PropertyRef = PropertyRef('profile.display_name')
+    first_name: PropertyRef = PropertyRef('profile.first_name')
+    last_name: PropertyRef = PropertyRef('profile.last_name')
+    profile_title: PropertyRef = PropertyRef('profile.title')
+    profile_phone: PropertyRef = PropertyRef('profile.phone')
+    email: PropertyRef = PropertyRef('profile.email')
```
Comment (ramonpetgrave64) [inline cartography/models/slack/user.py:25]: Could be worth an extra_index=True

File: tests/integration/cartography/intel/slack/test_team.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        neo4j_session,
+        slack_client,
+        SLACK_TEAM_ID,
+        TEST_UPDATE_TAG,
+        {"UPDATE_TAG": TEST_UPDATE_TAG, "TEAM_ID": SLACK_TEAM_ID},
+    )
+
+    # Assert Team exists
+    expected_nodes = {
+        (SLACK_TEAM_ID, 'Lyft OSS'),
```
Comment (ramonpetgrave64) [inline tests/integration/cartography/intel/slack/test_team.py:32]: ❤️

**Review by ramonpetgrave64 — CHANGES_REQUESTED** [review]
This is awesome! 
Just a few small things to change

**Review by ramonpetgrave64 — CHANGES_REQUESTED** [review]
This is awesome! 
Just a few small things to change

File: cartography/intel/slack/__init__.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+    common_job_parameters = {
+        "UPDATE_TAG": config.update_tag,
+    }
+
+    rate_limit_handler = RateLimitErrorRetryHandler(max_retry_count=1)
+    slack_client = WebClient(token=config.slack_token)
+    slack_client.retry_handlers.append(rate_limit_handler)
+
+    for team_id in config.slack_teams.split(','):
```
[the remark this answers is not in the document]
Comment (resilience-jychp) [reply cartography/intel/slack/__init__.py:41]: I will explore this.

File: cartography/models/slack/user.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+class SlackTeamToSlackUserRelProperties(CartographyRelProperties):
+    lastupdated: PropertyRef = PropertyRef('lastupdated', set_in_kwargs=True)
+
+
+@dataclass(frozen=True)
+# (:SlackTeam)<-[:RESOURCE]-(:SlackUser)
+class SlackTeamToUserRel(CartographyRelSchema):
+    target_node_label: str = 'SlackTeam'
+    target_node_matcher: TargetNodeMatcher = make_target_node_matcher(
+        {'id': PropertyRef('TEAM_ID', set_in_kwargs=True)},
```
[the remark this answers is not in the document]
Comment (resilience-jychp) [reply cartography/models/slack/user.py:63]: Yes but you can create  a sub-resource relationship on with in_kwargs ...
See https://lyftoss.slack.com/archives/CTZUQL0KX/p1685344006183709?thread_ts=1685344006.183709&cid=CTZUQL0KX

File: cartography/intel/slack/team.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+def sync(
+    neo4j_session: neo4j.Session,
+    slack_client: WebClient,
+    team_id: str,
+    update_tag: int,
+    common_job_parameters: Dict[str, Any],
+) -> None:
+    team = get(slack_client, team_id)
+    load_team(neo4j_session, team, update_tag)
+
```
[the remark this answers is not in the document]
Comment (resilience-jychp) [reply cartography/intel/slack/team.py:25]: Can not create an automatic cleanup, but clearly I can create a manual one.

File: cartography/models/slack/group.py (hunk 2 above)
Comment (resilience-jychp) [reply cartography/models/slack/group.py:18]: Clearly.
Worth it to add "Check extra index requirements" to your PR template ;)

File: cartography/intel/slack/users.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    team_id: str,
+    update_tag: int,
+) -> None:
+
+    load(
+        neo4j_session,
+        SlackUserSchema(),
+        data,
+        lastupdated=update_tag,
+        TEAM_ID=team_id,
```
[the remark this answers is not in the document]
Comment (resilience-jychp) [reply cartography/intel/slack/users.py:55]: @ramonpetgrave64 yes but needed for subressource matching

See https://lyftoss.slack.com/archives/CTZUQL0KX/p1685344006183709?thread_ts=1685344006.183709&cid=CTZUQL0KX

File: cartography/intel/slack/channels.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+
+@timeit
+def get(slack_client: WebClient, team_id: str, cursor: Optional[str] = None) -> List[Dict[str, Any]]:
+    channels: List[Dict[str, Any]] = []
+    channels_info = slack_client.conversations_list(cursor=cursor, team_id=team_id)
+    for m in channels_info['channels']:
+        channels.append(m)
+        if m['is_archived']:
+            channels.append(m)
```
[the remark this answers is not in the document]
Comment (resilience-jychp) [reply cartography/intel/slack/channels.py:38]: No, clearly an artefact (that will not lead to a bug but unecessary processing).

File: cartography/models/slack/channels.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+
+@dataclass(frozen=True)
+# (:SlackTeam)<-[:RESOURCE]-(:SlackChannel)
+class SlackTeamToChannelRel(CartographyRelSchema):
+    target_node_label: str = 'SlackTeam'
+    target_node_matcher: TargetNodeMatcher = make_target_node_matcher(
+        {'id': PropertyRef('TEAM_ID', set_in_kwargs=True)},
+    )
+    direction: LinkDirection = LinkDirection.OUTWARD
```
Comment (achantavy) [inline cartography/models/slack/channels.py:71]:
```suggestion
    direction: LinkDirection = LinkDirection.INWARD
```

File: cartography/models/slack/channels.py [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+    properties: SlackChannelToSlackUserRelProperties = SlackChannelToSlackUserRelProperties()
+
+
+@dataclass(frozen=True)
+class SlackTeamToSlackChannelRelProperties(CartographyRelProperties):
+    lastupdated: PropertyRef = PropertyRef('lastupdated', set_in_kwargs=True)
+
+
+@dataclass(frozen=True)
+# (:SlackTeam)<-[:RESOURCE]-(:SlackChannel)
```
Comment (achantavy) [inline cartography/models/slack/channels.py:65]:
```suggestion
# (:SlackTeam)-[:RESOURCE]->(:SlackChannel)
```
In AWS, we have the RESOURCE relationship point _from the account to the resource_ like `(:AWSAccount)-[:RESOURCE]->(:EC2Instance)`, so we should follow that pattern here with Slack Team going from the Team to the Channel
https://github.com/lyft/cartography/blob/0c852d26ec32e7dedda28d260290023ea98f3b7e/cartography/models/aws/ec2/instances.py#L53

File: cartography/models/slack/group.py [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+    properties: SlackGroupToCreatorRelProperties = SlackGroupToCreatorRelProperties()
+
+
+@dataclass(frozen=True)
+class SlackTeamToSlackGroupRelProperties(CartographyRelProperties):
+    lastupdated: PropertyRef = PropertyRef('lastupdated', set_in_kwargs=True)
+
+
+@dataclass(frozen=True)
+# (:SlackTeam)<-[:RESOURCE]-(:SlackGroup)
```
Comment (achantavy) [inline cartography/models/slack/group.py:72]:
```suggestion
# (:SlackTeam)-[:RESOURCE]->(:SlackGroup)
```

File: cartography/models/slack/group.py [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+class SlackTeamToGroupRel(CartographyRelSchema):
+    target_node_label: str = 'SlackTeam'
+    target_node_matcher: TargetNodeMatcher = make_target_node_matcher(
+        {'id': PropertyRef('TEAM_ID', set_in_kwargs=True)},
+    )
+    direction: LinkDirection = LinkDirection.OUTWARD
```
Comment (achantavy) [inline cartography/models/slack/group.py:78]:
```suggestion
    direction: LinkDirection = LinkDirection.INWARD
```

File: cartography/models/slack/user.py [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+    )
+    direction: LinkDirection = LinkDirection.OUTWARD
```
Comment (achantavy) [inline cartography/models/slack/user.py:66]:
```suggestion
    direction: LinkDirection = LinkDirection.INWARD
```

**Review by achantavy — CHANGES_REQUESTED** [review]
Overall it looks good to me, left comments about relationship direction for `RESOURCE`

File: cartography/intel/slack/channels.py [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+                channel_m = m.copy()
+                channel_m['member_id'] = i
+                channels.append(channel_m)
+    next_cursor = channels_info.get('response_metadata', {}).get('next_cursor', '')
+    if next_cursor != '':
+        channels.extend(get(slack_client, team_id, cursor=next_cursor))
+    return channels
+
+
+def _get_membership(slack_client: WebClient, slack_channel: str, cursor: Optional[str] = None) -> List[str]:
```
[the remark this answers is not in the document]
Comment (resilience-jychp) [reply cartography/intel/slack/channels.py:49]: @achantavy only need it for edge case

May be not pulling membership by default but allowing it with an argument like `--slack_get_memberships` can be a good alternative, what do you think ?

File: docs/root/modules/slack/config.md [hunk 17]
```diff
@@ -0,0 +1,25 @@
+## Slack Configuration
+
+.. _slack_config:
+
+Follow these steps to analyze Slack objects with Cartography.
+
+1. Create a Slack integration
```
[the remark this answers is not in the document]
Comment (resilience-jychp) [reply docs/root/modules/slack/config.md:7]: No only private channels where bot is invited in.
Did not find a way to list all private channels :(

**Comment by resilience-jychp** [comment]
@achantavy @ramonpetgrave64 Two remaining points before we can proceed:

1. Regarding dynamic team pulling, I suggest using teams IDs provided as arguments. This is because we cannot define sub-resource links on dynamically added items.

2. Regarding manual cleanup, do you have any guidance on how to perform it?

**Pull request closed without merging**